sql/analyze.go

```diff
@@ -0,0 +1,928 @@
+		return splitAndExpr(t.Right, splitAndExpr(t.Left, exprs))
+	}
+	return append(exprs, e)
+}
+
+// simplifyExpr transforms an expression such that it contains only expressions
+// involving qvalues that can be used for index selection. If an expression is
+// encountered that cannot be used for index selection (e.g. "func(val)") that
+// part of the expression tree is considered to evaluate to true, possibly
+// rendering the entire expression as true. Additionally, various
+// normalizations are performed on comparison expressions. For example:
+//
+//   (a < 1 AND a < 2)  -> (a < 1)
+//   (a < 1 AND a > 2)  -> false
+//   (a > 1 OR a < 2)   -> true
+//   (a > 1 OR func(b)) -> true
+//
+// Note that simplification is not normalization. Normalization as performed by
+// parser.NormalizeExpr returns an expression that is equivalent to the
+// original. Simplification can return an expression with parts of the
+// expression tree stripped out.
+func simplifyExpr(e parser.Expr) parser.Expr {
+	switch t := e.(type) {
+	case *parser.NotExpr:
+		return simplifyNotExpr(t)
+	case *parser.AndExpr:
+		return simplifyAndExpr(t)
+	case *parser.OrExpr:
+		return simplifyOrExpr(t)
+	case *parser.ComparisonExpr:
+		return simplifyComparisonExpr(t)
+	case *qvalue:
+		return e
+	}
+	// We don't know how to simplify expressions that fall through to here, so
+	// consider this part of the expression true.
+	return parser.DBool(true)
+}
+
+func simplifyNotExpr(n *parser.NotExpr) parser.Expr {
+	switch t := n.Expr.(type) {
+	case *parser.ComparisonExpr:
+		op := t.Operator
+		switch op {
+		case parser.EQ:
+			op = parser.NE
+		case parser.NE:
+			op = parser.EQ
+		case parser.GT:
+			op = parser.LT
+		case parser.GE:
+			op = parser.LE
+		case parser.LT:
+			op = parser.GT
+		case parser.LE:
+			op = parser.GE
+		case parser.Like:
+			op = parser.NotLike
+		case parser.NotLike:
+			op = parser.Like
+		case parser.SimilarTo:
+			op = parser.NotSimilarTo
+		case parser.NotSimilarTo:
+			op = parser.SimilarTo
+		default:
+			return parser.DBool(true)
+		}
+		return simplifyExpr(&parser.ComparisonExpr{
+			Operator: op,
+			Left:     t.Left,
+			Right:    t.Right,
+		})
+
+	case *parser.AndExpr:
+		// De Morgan's Law: (a AND b) -> (NOT a) OR (NOT b)
+		return simplifyExpr(&parser.OrExpr{
+			Left:  &parser.NotExpr{Expr: t.Left},
+			Right: &parser.NotExpr{Expr: t.Right},
+		})
+
+	case *parser.OrExpr:
+		// De Morgan's Law: NOT (a OR b) -> (NOT a) AND (NOT b)
+		return simplifyExpr(&parser.AndExpr{
+			Left:  &parser.NotExpr{Expr: t.Left},
+			Right: &parser.NotExpr{Expr: t.Right},
+		})
+	}
+	return parser.DBool(true)
+}
+
+func simplifyAndExpr(n *parser.AndExpr) parser.Expr {
+	// a AND b AND c AND d -> [a, b, c, d]
+	exprs := splitAndExpr(n, nil)
+	for i := range exprs {
+		exprs[i] = simplifyExpr(exprs[i])
+		if d, ok := exprs[i].(parser.DBool); ok && !bool(d) {
+			return d
+		}
+	}
+
+	// Loop over the expressions looking for simplifications.
```

Comment:
This is an `O(n^2)` loop. We could do better if we sorted it based on the variables in the expressions. Doesn't seem important to optimize right now. I'm adding a TODO.